adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1148,14 +1149,92 @@ protected final void onCommandReceived(
         final String commandRequestId = (command.isOneWay() ? "" : command.getRequestId());
         final String topic = String.format("%s/%s/%s/%s/%s/%s", subscription.getEndpoint(), tenantId, deviceId,
                 subscription.getRequestPart(), commandRequestId, command.getName());
-        endpoint.publish(topic, command.getPayload(), qos, false, false);
+        LOG.debug("Publishing command to device [tenant-id: {}, device-id: {}, MQTT client-id: {}, QoS: {}]",
+                        subscription.getTenant(), subscription.getDeviceId(), subscription.getClientId(), subscription.getQos());
+        commandContext.getCurrentSpan().log(String.format("Publishing command to device [tenant-id: %s, device-id: %s, MQTT client-id: %s, QoS: %s]",
+                subscription.getTenant(), subscription.getDeviceId(), subscription.getClientId(), subscription.getQos()));
+
+        if (MqttQoS.AT_LEAST_ONCE.equals(subscription.getQos())) {
+            registerPublishAcknowledgeHandler(endpoint, subscription);
```

Comment:
My understanding of vertx-mqtt is that the PUBACK handler is registered once per endpoint only, or is there a particular reason why we need to do this *for each* command? What if an application sends two commands in direct succession, i.e. without waiting for the first command's response to arrive before sending the second one?

Reply:
It makes sense to register once per endpoint. I will push the changed version.